File: apps/web/src/core/domain/money.ts
/**
 * Dinheiro no LUC é exato e em centavos (invariante #6 do CONTEXT.md): inteiro,
 * BRL, nunca ponto flutuante. Esta é a única forma de virar texto para a UI.
 */

/** Formata centavos (inteiro) como BRL — "R$ 1.234,56" (ponto de milhar, vírgula decimal). */
export function formatBRL(cents: number): string {
  if (!Number.isInteger(cents)) {
    throw new Error(`valor monetário deve ser inteiro em centavos, recebido: ${cents}`)
  }

  const negativo = cents < 0
  const abs = Math.abs(cents)
  const reais = Math.floor(abs / 100)
  const centavos = abs % 100

  const reaisComMilhar = String(reais).replace(/\B(?=(\d{3})+(?!\d))/g, ".")
  const centavosFmt = String(centavos).padStart(2, "0")

  return `${negativo ? "-" : ""}R$ ${reaisComMilhar},${centavosFmt}`
}

/**
 * Lê um valor BRL digitado (a borda da baixa) em centavos inteiros — `null` se
 * não for dinheiro válido e positivo. Aceita o formato brasileiro com milhar
 * ("1.234,56" · "1.500"), a vírgula decimal sem milhar ("19,99") e o ponto
 * decimal do teclado numérico ("1234.56" · "129.90"), além do inteiro de reais
 * ("1500" → R$ 1.500,00). Recusa mais de 2 casas decimais, sinal negativo (uma
 * baixa é positiva) e caractere fora do esperado. Sem ponto flutuante: monta os
 * centavos a partir das partes inteiras (#6).
 */
export function parseCentavos(texto: string): number | null {
  const limpo = texto.trim().replace(/^R\$/i, "").replace(/\s/g, "")
  if (limpo === "" || !/^[\d.,]+$/.test(limpo)) return null

  const normal = limpo.includes(",")
    ? // Vírgula é o decimal; pontos são de milhar.
      limpo.replace(/\./g, "").replace(",", ".")
    : normalizarSoComPontos(limpo)
  if (normal === null || !/^\d+(\.\d{1,2})?$/.test(normal)) return null

  const [reais, frac = ""] = normal.split(".")
  const centavos = Number(reais) * 100 + Number(frac.padEnd(2, "0"))
  if (!Number.isSafeInteger(centavos) || centavos <= 0) return null
  return centavos
}

/**
 * Resolve a ambiguidade do ponto sem vírgula: dinheiro tem 2 casas, então um
 * último grupo de 3 dígitos é milhar ("1.500" → "1500"); de 1–2 dígitos é
 * decimal ("129.90" → "129.90", com pontos anteriores como milhar). `null` se a
 * forma não couber em nenhum dos dois (ex.: 4+ dígitos após o ponto).
 */
function normalizarSoComPontos(limpo: string): string | null {
  if (!limpo.includes(".")) return limpo
  const partes = limpo.split(".")
  const ultima = partes[partes.length - 1]
  if (ultima.length === 3) return partes.join("")
  if (ultima.length === 1 || ultima.length === 2) {
    const dec = partes.pop() as string
    return `${partes.join("")}.${dec}`
  }
  return null
}

/**
 * Projeta centavos no texto que o campo de valor edita ("1234,56" — vírgula
 * decimal, sem milhar), para pré-preencher a baixa e a edição. Round-trip com
 * `parseCentavos`. Distinto de `formatBRL`, que é para *exibir* ("R$ 1.234,56").
 */
export function centavosParaCampo(cents: number): string {
  if (!Number.isInteger(cents)) {
    throw new Error(`valor monetário deve ser inteiro em centavos, recebido: ${cents}`)
  }
  const reais = Math.trunc(cents / 100)
  const centavos = Math.abs(cents % 100)
  return `${reais},${String(centavos).padStart(2, "0")}`
}

File: apps/api/src/luc_api/shared/domain/money.py
"""Dinheiro no LUC é exato e em centavos (invariante #6): inteiro, BRL, nunca float.

É a única forma de virar valor monetário em texto para a UI.
"""

import re

_TAM_GRUPO_MILHAR = 3  # dinheiro tem 2 casas → grupo final de 3 dígitos é milhar
_MIN_CASAS_DECIMAIS = 1
_MAX_CASAS_DECIMAIS = 2

_SO_DIGITOS_PONTO_VIRGULA = re.compile(r"[\d.,]+")
_VALOR_NORMALIZADO = re.compile(r"\d+(\.\d{1,2})?")


def _garantir_inteiro(cents: object) -> None:
    """Rejeita valor monetário que não seja inteiro em centavos (invariante #6)."""
    if not isinstance(cents, int) or isinstance(cents, bool):
        raise ValueError(f"valor monetário deve ser inteiro em centavos, recebido: {cents}")


def _com_milhar(reais: int) -> str:
    """Insere o ponto de milhar num inteiro de reais (1000000 → "1.000.000")."""
    return f"{reais:,}".replace(",", ".")


def format_brl(cents: int) -> str:
    """Formata centavos (inteiro) como BRL — "R$ 1.234,56" (milhar com ponto)."""
    _garantir_inteiro(cents)
    negativo = cents < 0
    abs_cents = abs(cents)
    reais = abs_cents // 100
    centavos = abs_cents % 100
    return f"{'-' if negativo else ''}R$ {_com_milhar(reais)},{centavos:02d}"


def format_brl_sem_centavos(cents: int) -> str:
    """Formata centavos como BRL sem casa decimal — "R$ 1.240" (estimativa, ≈).

    Arredonda ao real mais próximo por aritmética inteira (sem float): uma média
    não finge a precisão de um fato. Fatos exatos continuam em `format_brl`.
    """
    _garantir_inteiro(cents)
    negativo = cents < 0
    reais = (abs(cents) + 50) // 100
    return f"{'-' if negativo else ''}R$ {_com_milhar(reais)}"


def parse_centavos(texto: str) -> int | None:
    """Lê um valor BRL digitado em centavos inteiros — `None` se inválido/não-positivo.

    Aceita o formato brasileiro com milhar ("1.234,56" · "1.500"), a vírgula
    decimal sem milhar ("19,99") e o ponto decimal do teclado numérico
    ("1234.56"), além do inteiro de reais ("1500" → R$ 1.500,00). Recusa mais de
    2 casas decimais, sinal negativo e caractere fora do esperado. Sem float:
    monta os centavos a partir das partes inteiras.
    """
    limpo = re.sub(r"^R\$", "", texto.strip(), flags=re.IGNORECASE)
    limpo = re.sub(r"\s", "", limpo)
    if limpo == "" or not _SO_DIGITOS_PONTO_VIRGULA.fullmatch(limpo):
        return None

    if "," in limpo:
        # Vírgula é o decimal; pontos são de milhar.
        normal = limpo.replace(".", "").replace(",", ".")
    else:
        normal = _normalizar_so_com_pontos(limpo)
    if normal is None or not _VALOR_NORMALIZADO.fullmatch(normal):
        return None

    reais, _, frac = normal.partition(".")
    centavos = int(reais) * 100 + int(frac.ljust(2, "0"))
    if centavos <= 0:
        return None
    return centavos


def _normalizar_so_com_pontos(limpo: str) -> str | None:
    """Resolve a ambiguidade do ponto sem vírgula: milhar (grupo final de 3) vs decimal.

    "1.500" → "1500" (milhar); "129.90" → "129.90" (decimal, com pontos
    anteriores como milhar). `None` se a forma não couber em nenhum dos dois.
    """
    if "." not in limpo:
        return limpo
    partes = limpo.split(".")
    tam_ultima = len(partes[-1])
    if tam_ultima == _TAM_GRUPO_MILHAR:
        return "".join(partes)
    if _MIN_CASAS_DECIMAIS <= tam_ultima <= _MAX_CASAS_DECIMAIS:
        dec = partes.pop()
        return f"{''.join(partes)}.{dec}"
    return None


def centavos_para_campo(cents: int) -> str:
    """Projeta centavos no texto que o campo de valor edita ("1234,56", sem milhar).

    Round-trip com `parse_centavos`. Distinto de `format_brl`, que é para exibir.
    """
    _garantir_inteiro(cents)
    reais = abs(cents) // 100
    centavos = abs(cents) % 100
    return f"{'-' if cents < 0 else ''}{reais},{centavos:02d}"
